phoenix-core/src/main/java/org/apache/phoenix/coprocessor/PhoenixTTLRegionObserver.java

```diff
@@ -0,0 +1,368 @@
+
+    @Override public void start(CoprocessorEnvironment e) throws IOException {
+        super.start(e);
+        metricSource = MetricsPhoenixCoprocessorSourceFactory.getInstance().getPhoenixTTLSource();
+    }
+
+    @Override public void stop(CoprocessorEnvironment e) throws IOException {
+        super.stop(e);
+    }
+
+    @Override
+    public RegionScanner postScannerOpen(ObserverContext<RegionCoprocessorEnvironment> c, Scan scan,
+            RegionScanner s) throws IOException {
+
+        if (!ScanUtil.isMaskTTLExpiredRows(scan) && !ScanUtil.isDeleteTTLExpiredRows(scan)) {
+            return s;
+        }
+
+        if (ScanUtil.isMaskTTLExpiredRows(scan)) {
+            metricSource.incrementMaskExpiredRequestCount();
+        }
+
+        if (ScanUtil.isDeleteTTLExpiredRows(scan)) {
+            metricSource.incrementDeleteExpiredRequestCount();
+        }
+
+        scan.setAttribute(PhoenixTTLRegionScanner.MASK_PHOENIX_TTL_EXPIRED_REQUEST_ID_ATTR,
+                Bytes.toBytes(metricSource.getMaskExpiredRequestCount()));
+
+        LOG.debug(String.format(
+                "********** PHOENIX-TTL: PhoenixTTLRegionObserver::postScannerOpen TTL for table = "
+                        + "[%s], scan = [%s], PHOENIX_TTL = %d ***************, numRequests=%d",
+                s.getRegionInfo().getTable().getNameAsString(), scan.toJSON(Integer.MAX_VALUE),
+                ScanUtil.getPhoenixTTL(scan), metricSource.getMaskExpiredRequestCount()));
+        return new PhoenixTTLRegionScanner(c.getEnvironment(), scan, s);
+    }
+
+    /**
+     * A region scanner that checks the TTL expiration of rows
+     */
+    private static class PhoenixTTLRegionScanner implements RegionScanner {
+        private static final String MASK_PHOENIX_TTL_EXPIRED_REQUEST_ID_ATTR =
+                "MASK_PHOENIX_TTL_EXPIRED_REQUEST_ID";
+
+        private final RegionScanner scanner;
+        private final Scan scan;
+        private final byte[] emptyCF;
+        private final byte[] emptyCQ;
+        private final Region region;
+        private final long minTimestamp;
+        private final long maxTimestamp;
+        private final long now;
+        private final boolean deleteIfExpired;
+        private final boolean maskIfExpired;
+        private final long requestId;
+        private long numRowsExpired;
+        private long numRowsScanned;
+        private long numRowsDeleted;
+        private boolean reported = false;
+
+        public PhoenixTTLRegionScanner(RegionCoprocessorEnvironment env, Scan scan,
+                RegionScanner scanner) throws IOException {
+            this.scan = scan;
+            this.scanner = scanner;
+            byte[] requestIdBytes = scan.getAttribute(MASK_PHOENIX_TTL_EXPIRED_REQUEST_ID_ATTR);
+            this.requestId = Bytes.toLong(requestIdBytes);
+
+            deleteIfExpired = ScanUtil.isDeleteTTLExpiredRows(scan);
+            maskIfExpired = !deleteIfExpired && ScanUtil.isMaskTTLExpiredRows(scan);
```

Comment:
Are there any checks to avoid a case where both of these attributes are set? Can something screw up if that does happen?